website/blog/2024-12-04-from-natural-language-to-sql/index.md

```diff
@@ -0,0 +1,1067 @@
+---
+title: "From Natural Language to SQL: Building and Tracking a Multi-Lingual Query Engine"
+description: MLflow Models from Code and MLflow Tracing applied to AI Workflows
+tags:
+  - pyfunc
+  - mlflow
+  - sql-generator
+  - models-from-code
+  - tracing
+slug: from-natural-language-to-sql
+authors:
+  - hugo-carvalho
+  - joana-ferreira
+  - rahul-pandey
+thumbnail: img/blog/from-natural-language-to-sql.png
+---
+If you're looking to build a Multilingual Query Engine that combines Natural Language-to-SQL generation with query execution while fully leveraging MLflow’s features, this blog post is your guide. We’ll explore how to leverage **MLflow Models from Code** to treat AI workflows as traditional ML models, enabling seamless tracking, versioning, and deployment across diverse serving infrastructures. Additionally, we’ll dive into MLflow’s **Tracing** feature, which enhances observability by tracking inputs, outputs, and metadata at every intermediate step of the AI workflow.
+
+<!-- truncate -->
+
+## Introduction
+
+SQL is an essential skill needed for managing and accessing data in relational databases and many users fail to possess the necessary skills that are required to write correct SQL queries which can hinder them from utilizing data effectively. NL2SQL or Natural Language to SQL systems help in solving this problem by providing a translation from Natural Language to SQL commands allowing non-technical people to interact with data: users can just ask questions in a natural language they are comfortable speaking and these systems will assist them in returning the appropriate information.
+
+However, there are also a number of problems that remain when creating an NL2SQL system like semantic ambiguity, schema mapping or error handling and user feedback. Therefore, it is very important that while building such systems, we must put some guardrails instead of completely relying on LLM.
+
+In this blog post, we’ll walk you through the process of building and managing the lifecycle of a Multilingual Query Engine, encompassing both Natural Language to SQL generation and query execution.
```

Comment:
I don't see much about lifecycle management in the post (it mentions lifecycle management a few times, but doesn't show much about how to use MLflow for that purpose). You might de-emphasize that but show a little more about how MLflow tracing gives visibility into the many different components of a setup like this.

I think it's really cool, looking at the gif at the end, how we can see the different nodes laid out in the article reflected in the trace. It might be interesting to show screenshots of that for each section, or at least call that out more clearly in that section—specifically, that the final graph can be a bit of a black box, might be challenging to debug, to figure out what is happening at each step, but tracing gives really clear visibility into that with the one line of code.